tests/test_qp.py

```diff
@@ -139,138 +177,253 @@ def test_create_qp_ex_no_attr_connected(self):
+                    try:
+                        qp = QP(ctx, qia)
+                    except PyverbsRDMAError as ex:
+                        if ex.error_code == errno.EOPNOTSUPP:
+                            raise unittest.SkipTest('Create UC QP with extended attrs is not supported')
+                        raise ex
+                    qp_attr, qp_init_attr = qp.query(e.IBV_QP_STATE | e.IBV_QP_CAP)
+                    verify_qp_attrs(caps, e.IBV_QPS_RESET, qp_init_attr, qp_attr)
+                    # Legacy QP
+                    qia = get_qp_init_attr(cq, attr)
+                    qia.qp_type = e.IBV_QPT_UC
+                    caps = qia.cap
+                    try:
+                        qp = QP(pd, qia)
+                    except PyverbsRDMAError as ex:
+                        if ex.error_code == errno.EOPNOTSUPP:
+                            raise unittest.SkipTest('Create UC QP is not supported')
+                        raise ex
+                    qp_attr, qp_init_attr = qp.query(e.IBV_QP_STATE | e.IBV_QP_CAP)
+                    verify_qp_attrs(caps, e.IBV_QPS_RESET, qp_init_attr, qp_attr)
+
+    def test_query_ud_qp(self):
+        """
+        Queries an UD QP after creation. Verifies that its properties are as expected.
+        """
+        for ctx, attr, attr_ex in self.devices:
+            with PD(ctx) as pd:
+                with CQ(ctx, 100, None, None, 0) as cq:
+                    # Extended QP
+                    qia = get_qp_init_attr_ex(cq, pd, attr, attr_ex, e.IBV_QPT_UD)
+                    caps = qia.cap # Save them to verify values later
+                    try:
+                        qp = QP(ctx, qia)
+                    except PyverbsRDMAError as ex:
+                        if ex.error_code == errno.EOPNOTSUPP:
+                            raise unittest.SkipTest('Create UD QP with extended attrs is not supported')
+                        raise ex
+                    qp_attr, qp_init_attr = qp.query(e.IBV_QP_STATE | e.IBV_QP_CAP)
+                    verify_qp_attrs(caps, e.IBV_QPS_RESET, qp_init_attr, qp_attr)
+                    # Legacy QP
+                    qia = get_qp_init_attr(cq, attr)
+                    qia.qp_type = e.IBV_QPT_UD
+                    caps = qia.cap
+                    try:
+                        qp = QP(pd, qia)
+                    except PyverbsRDMAError as ex:
+                        if ex.error_code == errno.EOPNOTSUPP:
+                            raise unittest.SkipTest('Create UD QP is not supported')
+                        raise ex
+                    qp_attr, qp_init_attr = qp.query(e.IBV_QP_STATE | e.IBV_QP_CAP)
+                    verify_qp_attrs(caps, e.IBV_QPS_RESET, qp_init_attr, qp_attr)
+
+    def test_query_raw_qp(self):
         """
-        Queries a QP after creation. Verifies that its properties are as
-        expected.
+        Queries an RAW Packet QP after creation. Verifies that its properties are as expected.
```

Comment:
keep docstrings cut at 80c (72c PEP recommendation)
(apply on the other relevant places)

Reply:
I'll fix it in v2.